File: tools/DumpInputCallers.py
# Ghidra headless script: dump decompiled callers of Windows mouse/input imports.
# @category ScrapMechanicVR

from ghidra.app.decompiler import DecompInterface
from java.io import FileWriter, BufferedWriter


names = [
    "GetRawInputData", "RegisterRawInputDevices", "GetCursorPos", "SetCursorPos",
    "PeekMessageW", "DispatchMessageW", "GetAsyncKeyState", "GetFocus"
]
arguments = getScriptArgs()
if len(arguments) != 1:
    raise RuntimeError("usage: DumpInputCallers.py OUTPUT.txt")

decompiler = DecompInterface()
decompiler.openProgram(currentProgram)
seen = set()
writer = BufferedWriter(FileWriter(arguments[0]))
try:
    table = currentProgram.getSymbolTable()
    references = currentProgram.getReferenceManager()
    functions = currentProgram.getFunctionManager()
    for name in names:
        writer.write("\n===== %s =====\n" % name)
        iterator = table.getSymbols(name)
        while iterator.hasNext():
            symbol = iterator.next()
            writer.write("symbol %s @ %s\n" % (symbol.getName(True), symbol.getAddress()))
            for reference in references.getReferencesTo(symbol.getAddress()):
                caller = functions.getFunctionContaining(reference.getFromAddress())
                if caller is None:
                    continue
                key = str(caller.getEntryPoint())
                writer.write("xref %s caller %s @ %s\n" %
                             (reference.getFromAddress(), caller.getName(), key))
                if key in seen:
                    continue
                seen.add(key)
                result = decompiler.decompileFunction(caller, 90, monitor)
                if result.decompileCompleted():
                    writer.write(result.getDecompiledFunction().getC())
                    writer.write("\n")
                else:
                    writer.write("decompile failed: %s\n" % result.getErrorMessage())
finally:
    writer.close()
    decompiler.dispose()

File: tools/DumpInputCallers.java
// Dump decompiled callers of Windows mouse/input imports.
// @category ScrapMechanicVR

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.util.HashSet;
import java.util.Set;

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Function;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceIterator;
import ghidra.program.model.symbol.Symbol;
import ghidra.program.model.symbol.SymbolIterator;

public class DumpInputCallers extends GhidraScript {
    @Override
    public void run() throws Exception {
        String[] arguments = getScriptArgs();
        if (arguments.length != 1) throw new IllegalArgumentException("usage: DumpInputCallers.java OUTPUT.txt");
        String[] names = {"GetRawInputData", "RegisterRawInputDevices", "GetCursorPos", "SetCursorPos",
                          "PeekMessageW", "DispatchMessageW", "GetAsyncKeyState", "GetFocus"};
        DecompInterface decompiler = new DecompInterface();
        decompiler.openProgram(currentProgram);
        Set<String> seen = new HashSet<>();
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(arguments[0]))) {
            for (String name : names) {
                writer.write("\n===== " + name + " =====\n");
                SymbolIterator symbols = currentProgram.getSymbolTable().getSymbols(name);
                while (symbols.hasNext()) {
                    Symbol symbol = symbols.next();
                    writer.write("symbol " + symbol.getName(true) + " @ " + symbol.getAddress() + "\n");
                    ReferenceIterator references = currentProgram.getReferenceManager().getReferencesTo(symbol.getAddress());
                    while (references.hasNext()) {
                        Reference reference = references.next();
                        Function caller = currentProgram.getFunctionManager().getFunctionContaining(reference.getFromAddress());
                        if (caller == null) continue;
                        String key = caller.getEntryPoint().toString();
                        writer.write("xref " + reference.getFromAddress() + " caller " + caller.getName() + " @ " + key + "\n");
                        if (!seen.add(key)) continue;
                        DecompileResults result = decompiler.decompileFunction(caller, 90, monitor);
                        if (result.decompileCompleted()) writer.write(result.getDecompiledFunction().getC() + "\n");
                        else writer.write("decompile failed: " + result.getErrorMessage() + "\n");
                    }
                }
            }
        } finally {
            decompiler.dispose();
        }
    }
}
